modules/map_preprocess.py

```diff
@@ -0,0 +1,234 @@
+"""
+This module contains all classes related to preprocessing of point data to
+be added to the occupancy map
+"""
+import time
+
+import numpy as np
+from scipy.spatial.transform import Rotation as R
+
+from modules import data_types, load_config
+from modules.async_message import AsyncMessageCallback
+
+
+class MapPreprocess:
+    """
+    This class takes a localised point cloud and prepares it to
+    be added to the global occupancy map
+    """
+    def __init__(self, config_file='conf/map.yaml'):
+        self.conf = load_config.from_file(config_file)
+        self.map_definition = load_config.conf_to_named_tuple(data_types.MapDefinition,
+                                                              self.conf.map.shape)
+
+        self.bins = None
+        self._initialise_bin_delimitations()
+
+        self.message_queue = AsyncMessageCallback(queue_size=1)
+        self._pose_data = None
+
+        self.last_time = time.time()
+
+    def depth_callback(self, data: data_types.Depth):
+        """
+        Receive incoming depth data
+        """
+        if self._pose_data is None:
+            return
+
+        self.message_queue.queue_message((data, self._pose_data))
+
+    def pose_callback(self, data: data_types.Pose):
+        """
+        Receive incoming pose data
+        """
+        self._pose_data = data
+
+    def process_incoming_data(self):
+        """
+        Process batches of local point clouds
+        """
+        data_in = self.message_queue.wait_for_message()
+
+        if data_in is None:
+            return None
+        
+        data, pose = data_in[1]
+
+        data = self._input_data_preprocess(data)
+        points, count = self._point_data_to_map(data, pose)
+
+        return data_types.MapPreProcessorOut(data_in[0], points, count)
+
+    def _input_data_preprocess(self, data):
+        """
+        Taken the raw data and convert to a standardised (Nx3) matrix of local
+        cartesian cooridantes. This function should be overriden by child class
+        """
+        raise NotImplementedError
+
+    def _point_data_to_map(self, point_cloud, pose):
+        """
+        Take coordinate array and convert to a set of map voxel coordinates and
+        a hit count for each voxel
+        """
+        map_points = self._local_to_global(point_cloud, pose)
+
+        if self.conf.map.enable_compression:
+            map_points, point_count = self._discretise_point_cloud(map_points)
+        else:
+            point_count = 1
+
+        return (map_points, point_count)
+
+    def _initialise_bin_delimitations(self):
```

Comment:
Minor point. I would almost always veer towards dependancy injection. So rather than this using internal state and setting a new internal state, it is probalby better to pass them in as parameters and return the value you are set. It prevents spooky behaviour at a distance and eases code flow/understanding.

Reply:
I started looking into this in the mapping bit, probably best to fix there where there's two modules that require the same grid definition to work